### PR TITLE
Use CXF to expose web services

### PR DESCRIPTION
**Pull request by daniel-dgi**
This solves jetty's multipart issues and lets us move up to karaf 4.0.2.

This address pulls #110 #95 and #57.

Sorry it's all jumbled up, but figured i'd do what I could while I had the stack ripped to pieces.

### CODE REVIEW

**Comment by daniel-dgi** [comment]
He'll be able to finish #83 now.

Crap... just realized i didn't run phpcs locallly.... If it pops, i'll patch it up tonight.

**Comment by DiegoPino** [comment]
@daniel-dgi, @ruebot, does this remake go into 2.x or 2.x-dev?

**Comment by DiegoPino** [comment]
Maybe we need a new version of hawtio? 1.4.56 ?

**Comment by DiegoPino** [comment]
@ruebot, if you got some time go through http://camel.apache.org/camel-2160-release.html, there is a Known Issues, dependencies upgrades, etc at the end. I'm sure there is some feature version conflicting somewhere.

**Comment by daniel-dgi** [comment]
bundle:list?  what's the results?  is the image service running?  a GET request with no UUID is not going to map to a function. 

can you try with http://127.0.0.1:8181/cxf/islandora/image/abc-123 and see what happens?  should be an exception thrown about not being able to locate a resource with uuid abc-123.

i'll vagrant destroy/up again and see what's happening.

**Comment by daniel-dgi** [comment]
Just re-vagrant'd and got this.  What's up with our install process where that isn't working for you?  :(

screen shot 2015-11-12 at 1 12 01 pm

Also, yes, I can confirm the hawtio problem.

**Comment by whikloj** [comment]
Pulled this down and built it, was able to add a collection and a basic image to the collection with auto-derivative creation. :+1:

**Comment by DiegoPino** [comment]
he! :+1:

**Comment by DiegoPino** [comment]
@ruebot, i will test tomorrow using your same versions to see what happens.

I think, maybe, we have a "recursive problem": since we are sharing some folders that are used to build the vagrant box and also symlinking, after a build and destroy, those folders keep the modifications made by the previous build, with stuff from the previous vagrant.. late night thinking(maybe even not  the problem you are seeing…). So in theory whipping the whole islandora-claw, cloning from git again and then doing a vagrant up should work.

After i got this weird drupal/drupal/drupal/drupal..etc (a very crazy recursive symlink) this week i think the problem could be there.

**Comment by daniel-dgi** [comment]
~ $ vagrant --version
Vagrant 1.7.2

**Comment by whikloj** [comment]
VirtualBox : 5.0.6 r103037
vagrant : 1.7.4
Ubuntu box : 14.04

Maybe because I haven't updated the basebox?

**Comment by daniel-dgi** [comment]
fyi, it's `vagrant box update` for those who wanna try.  i'm doing it now and seeing what happens with a fresh install.

File: camel/services/basic-image-service/pom.xml [hunk 1]
```diff
@@ -6,23 +6,24 @@
...
+    <fcrepo.camel.version>4.4.0</fcrepo.camel.version>
+    <cxf.version>3.1.4</cxf.version>
     <slf4j.version>1.7.12</slf4j.version>
     <log4j.version>1.2.17</log4j.version>
     <compiler.plugin.version>2.5.1</compiler.plugin.version>
     <java.version>1.8</java.version>
     <resources.plugin.version>2.6</resources.plugin.version>
     <bundle.plugin.version>2.3.7</bundle.plugin.version>
-    <islandora.osgi.export.pkg>ca.islandora.camel.services.*</islandora.osgi.export.pkg>
+    <islandora.osgi.export.pkg>ca.islandora.camel.services*</islandora.osgi.export.pkg>
```
Comment (whikloj) [inline camel/services/basic-image-service/pom.xml]: This is not code review, but a question. Why do you export ```ca.islandora.camel.services*``` here but ```ca.islandora.camel.services.collection*``` in the collection service? I don't understand where the package names come from.

Comment (daniel-dgi) [inline camel/services/basic-image-service/pom.xml]: That's actually a good catch.  I should be exporting the main package for the project.  Should be ca.islandora.camel.image.basic*

**Comment by daniel-dgi** [comment]
Also good catch.  Won't stop the build, but will prevent the auto-redeploy workflow from happening.

**Comment by daniel-dgi** [comment]
Just did a clean install from a fresh clone with an updated base box and it's still ok for me :(

I'm going to push up changes for hawt and the watch scripts and the package exports soon.

**Comment by daniel-dgi** [comment]
@ruebot but did it start?  what is the bundle's status?  anything in the logs?  i see that too when i manually install the feature.  no problems on my end.  may be another red herring.

**Comment by daniel-dgi** [comment]
@ruebot well that's good.  so it's something with the install process that's bombing out, then.  anything earlier in your logs for when the feature install was first attempted?

**Comment by daniel-dgi** [comment]
ok, testing hawtio update to 1.4.56 as we speak.  i'll push it up along with the other changes once it's good.

File: camel/services/basic-image-service/pom.xml (hunk 1 above)
Comment (daniel-dgi) [inline camel/services/basic-image-service/pom.xml]: @whikloj to answer your question (b/c i just realized my response didn't address that at all), the export package name comes from what you want this OSGi bundle to provide if running in karaf.  it should be, by default, what packages your bundle provides.  that way other bundles that depend on this functionality can declare it in their import packages and OSGi's runtime checks will kick in and prevent deployment if the conditions aren't right.

admittedly, we're not really providing much, but it is definitely best we set this up right.

**Comment by DiegoPino** [comment]
@ruebot and @daniel-dgi 
Maybe it's not syncing because of

` --2015-11-13 15:29:35--  https://github.com/fcrepo4-labs/fcrepo-camel-toolbox/releases/download/fcrepo-camel-toolbox-4.4.0/fcrepo-camel-webapp-at-is-it-4.4.0.war
==> default: Resolving github.com (github.com)... 
==> default: 192.30.252.130
==> default: Connecting to github.com (github.com)|192.30.252.130|:443... 
==> default: connected.
==> default: HTTP request sent, awaiting response... 
==> default: 301 Moved Permanently
==> default: Location: https://github.com/fcrepo4-exts/fcrepo-camel-toolbox/releases/download/fcrepo-camel-toolbox-4.4.0/fcrepo-camel-webapp-at-is-it-4.4.0.war [following]
==> default: --2015-11-13 15:29:36--  https://github.com/fcrepo4-exts/fcrepo-camel-toolbox/releases/download/fcrepo-camel-toolbox-4.4.0/fcrepo-camel-webapp-at-is-it-4.4.0.war
==> default: Reusing existing connection to github.com:443.
==> default: HTTP request sent, awaiting response... 
==> default: 404 Not Found
==> default: 2015-11-13 15:29:36 ERROR 404: Not Found.`

**Comment by whikloj** [comment]
As diego said there is no WAR file for fcrepo-camel-toolbox version 4.4.0, so the download fails. Not sure why mine worked?

**Comment by daniel-dgi** [comment]
@ruebot ok, and so what happens if you try to install the feature?

@DiegoPino has nothing to do with sync.  would prevent triplestore indexing, though.

man, not getting anything like this on my end.  makes it awfully tough to diagnose.

**Comment by daniel-dgi** [comment]
well.... we can always wrap #83 in with this...  will take care of the war issue.

**Comment by DiegoPino** [comment]
@daniel-dgi, should affect if you have dependencies like in the image service, even when not used?  Testing is slow, sorry.

**Comment by daniel-dgi** [comment]
Looks like even newest hawtio karaf terminal plugin has issues with karaf 4.  the main guy from camel has reported it to the hawtio team:  https://github.com/hawtio/hawtio/issues/2028

i'll push up all the changes, and we'll have to set up a seperate ticket for hawtio.

**Comment by whikloj** [comment]
Ok so according to [FCREPO-1691](https://jira.duraspace.org/browse/FCREPO-1691) there will be no more war files created for fcrepo-camel-toolbox. 
So best to move the version back to 4.3.0 or build [fcrepo-camel-webapp](https://github.com/fcrepo4-labs/fcrepo-camel-webapp) from source.

**Comment by daniel-dgi** [comment]
or i'll just roll in #83... seriously.  this has already turned into a total cluster.  may as well just throw it in.

**Comment by daniel-dgi** [comment]
@ruebot lololololol.... pretty much.  happens sometimes.

**Comment by daniel-dgi** [comment]
At this point i'm wondering if it's just how we're declaring things in our features.xml files.  I just spun up a new instance and finally got a 'no service found' error.  Checked the logs and there's all this stuff about mustache and http not being found, when the bundles indeed exist and are enabled!

So maybe it's a timing issue with how features installs things.  Sadly, the documentation's not so great, so I'll be scouring an xsd to see if there's anything else I can add to the <feature> tags to control startup/dependency ordering.

**Comment by acoburn** [comment]
@daniel-dgi I haven't followed all of this thread, but in terms of feature dependencies, take a look at this: http://git.io/v4kqn

To use that attribute, you have to use v1.3.0 of the features namespace, which implies other changes to the features.xml file, such as removing any `resolver` attributes. I believe there are some other properties available for use on the `<feature>` elements, but you'll need to refer to the XSD document :-(

**Comment by DiegoPino** [comment]
@acoburn, if grouped under a common feature, would be the bundle order problem disappear? or i am talking non sense? Karaf does handle that well http://stackoverflow.com/questions/16418086/how-apache-karaf-sorts-bundles-to-install-and-start

**Comment by acoburn** [comment]
@DiegoPino you are correct, karaf does handle dependency resolution really well. I've only ever seen a need to explicitly state dependencies like this when modifying `$KARAF_HOME/etc/org.apache.karaf.features.repos.cfg` to define a set of features to install on the initial load of karaf. If you're manually installing features with `feature:install` this wouldn't be necessary.

**Comment by daniel-dgi** [comment]
@acoburn I'm thinking that it's an issue with us not having any import packages declared in the pom files for all the services.  I'll give that a whirl and see what happens.

I think Karaf is *starting* things in the right order, but we have no control over when things finish initializing.  I think the import package declaration should ameliorate this.

**Comment by daniel-dgi** [comment]
@acoburn well that failed miserably.  i'm onto trying the prerequisite attribute and we'll see how that goes.

**Comment by daniel-dgi** [comment]
OKAY.  Initial results with @acoburn's suggested fix are promising.  Feature installs are definitely happening with dependencies in the correct order now.  @ruebot @DiegoPino @whikloj can I get some verification on this?

**Comment by whikloj** [comment]
Building, but I am not starting with a fresh directory. Also I switched back to the fcrepo-camel-toolbox 4.3.0

**Comment by whikloj** [comment]
Sorry @ruebot , but it still works for me. @daniel-dgi, do you think I need to start with a fresh directory?
```
karaf@root()> feature:list | grep islandora
islandora-camel-component               | 0.0.1.SNAPSHOT   | x        | Started     | islandora-camel-component-0.0.1-SNAPSHOT     |
islandora-collection-service            | 0.0.1.SNAPSHOT   | x        | Started     | islandora-collection-service-0.0.1-SNAPSHOT  |
islandora-basic-image-service           | 0.0.1.SNAPSHOT   | x        | Started     | islandora-basic-image-service-0.0.1-SNAPSHOT |
islandora-sync-gateway                  | 0.0.1.SNAPSHOT   | x        | Started     | islandora-sync-gateway-0.0.1-SNAPSHOT        |
karaf@root()> bundle:list | grep islandora
129 | Active  |  80 | 0.0.1.SNAPSHOT | islandora-camel-component
185 | Active  |  80 | 0.0.1.SNAPSHOT | islandora-sync-gateway
258 | Active  |  80 | 0.0.1.SNAPSHOT | islandora-collection-service
259 | Active  |  80 | 0.0.1.SNAPSHOT | islandora-basic-image-service
karaf@root()> 
```

**Comment by daniel-dgi** [comment]
@whikloj  Every time I've tried with a fresh, new clone of the repo just in case and it works fine.  Maybe we should test on digital ocean. If it works there, I'd be inclined to suspect @ruebot's environment.  What are you using? I'm on a year old Mac book pro using Mavericks.

**Comment by whikloj** [comment]
@daniel-dgi mine is a 2 year old Mac book pro still on Mountain Lion.

**Comment by daniel-dgi** [comment]
@whikloj hrm... mac trend for the ok's.  with @nigelgbanks's report of bad install on ubunutu at #43 is in line with @ruebot's specs.  are you 14.04 @ruebot?

**Comment by DiegoPino** [comment]
@ruebot, @whikloj is having that problem right now.  Can you see if the camel-http feature is installed?

**Comment by daniel-dgi** [comment]
Check camel versions, too.  I think the fcrepo-camel-toolbox machinery is on 2.15.1

**Comment by daniel-dgi** [comment]
Hrm... so the issue definitely appears to be the install is dying when doing stuff with Karaf.  Maybe piping the file into Karaf isn't the best approach?  See https://github.com/fcrepo4-exts/fcrepo4-vagrant/blob/master/install_scripts/fedora_camel_toolbox.sh#L15.  Maybe using the -f flag will help.

It's awfully hard doing this blind.  I need to get my hands on a non-virtualized linux box.

**Comment by DiegoPino** [comment]
@daniel-dgi, if we instead of doing the piping, we define them in  etc/org.apache.karaf.features.cfg ?
For those who haven't read this one (not your case dani!), some documentation good to have here
https://karaf.apache.org/manual/latest/users-guide/provisioning.html

**Comment by whikloj** [comment]
This is where I am, I think part of it is we need fcrepo-camel which uses camel 2.15.1 and we are using 2.16.0 so although Karaf should handle this, it doesn't. Most of my failures seem to occur initializing basic-image-service and collection-service in version 2.15.1, even though the pom requires version 2.16 up to 3.

**Comment by daniel-dgi** [comment]
Need to do some try/catch action here https://github.com/Islandora-CLAW/CLAW/blob/7.x-2.x/camel/sync/src/main/resources/OSGI-INF/blueprint/eventGateway.xml#L169.  See https://github.com/Islandora-CLAW/CLAW/blob/7.x-2.x/camel/sync/src/main/resources/OSGI-INF/blueprint/eventGateway.xml#L81-L89 as an example of doing exception handling in the blueprint dsl.

**Comment by daniel-dgi** [comment]
Ok, so the problem here is that Fedora results are actually a graph of nodes, not always just the single node.  So sometimes, like when there's an fcr:metadata, there's actually two entries in the json array.  Looks like here there's only one.  So set it to [0] and see what else breaks?  We may be in a scenario where we need to normalize the results coming out of fedora to be consistent results (probably some trickery with the 'Prefer' header).  Ideally, we'd be parsing this stuff into a Graph object (assuming with Jena) and pulling out the node we want by id.  That's how we do it with Chullo, btw.

**Comment by daniel-dgi** [comment]
Ok, we've managed to get this installing for @ruebot and on digital ocean (also linux).  @nigelgbanks, can you give this a whirl on your machine and see if it installs for you?

**Comment by nigelgbanks** [comment]
I can't bring it up, though I suspect it has nothing to do with your pull request, and more so is a problem with vagrant itself, perhaps my install? I ran vagrant up twice and this was it's output each time:

[vagrant.txt](https://github.com/Islandora-CLAW/CLAW/files/41883/vagrant.txt)
[vagrant_2.txt](https://github.com/Islandora-CLAW/CLAW/files/41882/vagrant_2.txt)

**Pull request merged**